{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b38f30e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from PIL import Image\n",
    "import plotly.express as px\n",
    "import plotly.colors as pc\n",
    "from pdb import set_trace\n",
    "import re\n",
    "import pandas as pd\n",
    "import os\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9a372c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the proteomics data and only keep rows where gene has value for all samples\n",
    "PFP = '../Krugg_Breast_Cancer_Data/kr_pro_z.csv' # proteomics file path\n",
    "PD = pd.read_csv(PFP)\n",
    "\n",
    "PD.index = PD.loc[:,'Gene']\n",
    "PD = PD.loc[:,PD.columns!='Gene']\n",
    "\n",
    "DropRowIndices = [(sum(np.isnan(PD.loc[gene,:]))>0) for gene in PD.index]\n",
    "KeepRowIndices = [not x for x in DropRowIndices]\n",
    "PD = PD.loc[KeepRowIndices,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cdd1749c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the mRNA data and only keep rows where gene has value for all samples\n",
    "MFP = '../Krugg_Breast_Cancer_Data/kr_rna_z.csv' # mRNA file path\n",
    "MD = pd.read_csv(MFP)\n",
    "\n",
    "MD.index = MD.loc[:,'Gene']\n",
    "MD = MD.loc[:,MD.columns!='Gene']\n",
    "\n",
    "DropRowIndices = [(sum(np.isnan(MD.loc[gene,:]))>0) for gene in MD.index]\n",
    "KeepRowIndices = [not x for x in DropRowIndices]\n",
    "MD = MD.loc[KeepRowIndices,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a6e3ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the localization data\n",
    "LFP = '../MCF7_SubCell_Barcode.txt'\n",
    "LD = pd.read_csv(filepath_or_buffer=LFP,sep='\\t')\n",
    "LD.index = LD.loc[:,'Protein']\n",
    "LD = LD.loc[:,LD.columns!='Protein']\n",
    "\n",
    "#Not_Mito_Indices = LD.loc[:,'Neighborhood']!='Mitochondria'\n",
    "#LD.loc[Not_Mito_Indices,'Neighborhood'] = 'NotMitochondria'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0d5aae92",
   "metadata": {},
   "outputs": [],
   "source": [
    "NotUnclassInd = LD.loc[:,'Neighborhood'] != 'Unclassified'\n",
    "LD = LD.loc[NotUnclassInd,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00379601",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NotCytInd = LD.loc[:,'Neighborhood'] != 'Cytosol'\n",
    "#LD = LD.loc[NotCytInd,:]\n",
    "#NotUnclassInd = LD.loc[:,'Neighborhood'] != 'Unclassified'\n",
    "#D = LD.loc[NotUnclassInd,:]\n",
    "#NotMitoInd = LD.loc[:,'Neighborhood'] != 'Mitochondria'\n",
    "#LD = LD.loc[NotMitoInd,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d1aea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(LD.loc[:,'Neighborhood'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f9d0d0ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract Amino Acid Sequence Data from SwissProt FASTA File\n",
    "FASTA_FileName = '../SwissProtFASTA20230411.txt'\n",
    "Prots = pd.DataFrame()\n",
    "with open(FASTA_FileName) as FASTA_Data:\n",
    "    GeneCounter = 0\n",
    "    SequenceRecorded = False\n",
    "    Gene = ''\n",
    "    for line in FASTA_Data:\n",
    "        if line[0:3] == '>sp':\n",
    "            \n",
    "            if SequenceRecorded:\n",
    "                Prots.loc[Gene,'ProtSequence'] = ProtSequence\n",
    "            SequenceRecorded = False\n",
    "            \n",
    "            GeneSearch = re.search('GN=[a-zA-Z0-9\\-]* ',line)\n",
    "            \n",
    "            Gene = ''\n",
    "            if GeneSearch is not None:\n",
    "                SearchSpan = GeneSearch.span()\n",
    "                GeneBeginIndex = SearchSpan[0]+3\n",
    "                GeneEndIndex = SearchSpan[1]-1\n",
    "                Gene = line[GeneBeginIndex:GeneEndIndex]\n",
    "                ProtSequence = ''\n",
    "                GeneCounter = GeneCounter+1\n",
    "                \n",
    "        \n",
    "        if line.strip()[0].isupper() & line.strip()[-1].isupper():\n",
    "            ProtSequence = ProtSequence + line.strip()\n",
    "            SequenceRecorded = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "de845fd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5946"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only rows whose genes are represented in both data sets\n",
    "IntersectingGenes = [value for value in PD.index if ((value in MD.index) & (value in LD.index) & (value in Prots.index))]\n",
    "PD = PD.loc[IntersectingGenes,:]\n",
    "MD = MD.loc[IntersectingGenes,:]\n",
    "Prots = Prots.loc[IntersectingGenes,:]\n",
    "len(MD.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04b3953e",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(Prots.loc['BRK1','ProtSequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7c0e05c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "3000\n",
      "3100\n",
      "3200\n",
      "3300\n",
      "3400\n",
      "3500\n",
      "3600\n",
      "3700\n",
      "3800\n",
      "3900\n",
      "4000\n",
      "4100\n",
      "4200\n",
      "4300\n",
      "4400\n",
      "4500\n",
      "4600\n",
      "4700\n",
      "4800\n",
      "4900\n",
      "5000\n",
      "5100\n",
      "5200\n",
      "5300\n",
      "5400\n",
      "5500\n",
      "5600\n",
      "5700\n",
      "5800\n",
      "5900\n"
     ]
    }
   ],
   "source": [
    "# Save one square image per gene with the pixels arranged as:\n",
    "#    patient1  patient1  patient2 patient2  ...  patient14  patient14\n",
    "#     prot     mRNA       prot     mRNA           prot       mRNA\n",
    "#                                .\n",
    "#                                .              \n",
    "#                                .\n",
    "#    patient85  patient85  patient86  patient86  ...  patient98  patient98\n",
    "#      prot       mRNA       prot       mRNA            prot       mRNA\n",
    "\n",
    "i = 0\n",
    "for gene in PD.index:\n",
    "    GeneImages(PD,MD,gene)\n",
    "    if i%100==0:\n",
    "        print(i)\n",
    "    i = i+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "62fb77ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Gene Images\n",
    "def GeneImages(PD,MD,gene):\n",
    "    # Make an array for each gene where the protein data is followed by the mRNA value for every patient\n",
    "    # patient1  patient1  patient2 patient2  ...  patient98  patient98\n",
    "    #   prot     mRNA       prot     mRNA           prot       mRNA\n",
    "\n",
    "    # Each image will have room for 162 samples\n",
    "    n_sample_spaces = 162\n",
    "    \n",
    "    # The number of samples in the given data set\n",
    "    n_samples = len(PD.columns)\n",
    "    \n",
    "    # Include protein sequence info in the image\n",
    "    Add_AA_Sequence = True\n",
    "    \n",
    "    GeneArray = np.array([])\n",
    "    for i in np.arange(0,n_sample_spaces):\n",
    "        # if there is data for a sample, access it\n",
    "        if i<n_samples:\n",
    "            Array = np.array([PD.loc[gene,:][i],MD.loc[gene,:][i]])\n",
    "        # if a sample space needs to be written with blank data, do so\n",
    "        if i>=n_samples:\n",
    "            Array = np.array([float(\"NaN\"),float(\"NaN\")])\n",
    "\n",
    "        GeneArray = np.concatenate((GeneArray,Array))\n",
    "            \n",
    "    \n",
    "    # Specify whether the array should be arranged by sum of neighboring pairs from smallest to largest\n",
    "    ReorderBySum = False\n",
    "    if ReorderBySum:\n",
    "        GeneArray = Reorder(GeneArray)\n",
    "        \n",
    "    if Add_AA_Sequence:\n",
    "        AA_DF = pd.DataFrame()\n",
    "        AA_DF.index = np.array(['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'])\n",
    "        AA_DF.loc[:,'Number'] = [100,200,300,400,500,600,700,800,900,1000,1100,1200,1300,1400,1500,1600,1700,1800,1900,2000,2100,2200,2300,2400,2500,2600]\n",
    "        \n",
    "        AA_Sequence = Prots.loc[gene,'ProtSequence']\n",
    "        \n",
    "        limit = n_sample_spaces-n_samples\n",
    "        AA_counter = 0\n",
    "        for i in np.arange(n_samples,n_sample_spaces):\n",
    "            if AA_counter < len(AA_Sequence):\n",
    "                AA = AA_Sequence[AA_counter]\n",
    "                AA_Number = AA_DF.loc[AA,'Number']\n",
    "                GeneArray[2*i] = AA_Number\n",
    "                AA_counter = AA_counter+1\n",
    "                \n",
    "            if AA_counter < len(AA_Sequence):  \n",
    "                AA = AA_Sequence[AA_counter]\n",
    "                AA_Number = AA_DF.loc[AA,'Number']\n",
    "                GeneArray[2*i+1] = AA_Number\n",
    "                AA_counter = AA_counter+1\n",
    "                \n",
    "    Square_GeneArray = SquareArray(GeneArray)\n",
    "\n",
    "    # Get the RGB array\n",
    "    RGB = CreateImage(Square_GeneArray,AA_DF)\n",
    "    RGB = np.asarray(RGB,dtype=np.uint8)\n",
    "\n",
    "    # Make an image from the RGB array\n",
    "    image = Image.fromarray(RGB)\n",
    "    \n",
    "    # Get compartment\n",
    "    compartment = LD.loc[gene,'Neighborhood']\n",
    "    \n",
    "    # Display and save the image\n",
    "    image.save('../Gene_Images/Krug_18by18Grid_Seq/'+gene+'.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9c6544ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to select a color from a pre-defined continuous pallet.\n",
    "def GetColor(z,AA_DF):\n",
    "    \n",
    "    # If there is a data point, find the color corresponding to the data point\n",
    "    if (not np.isnan(z)) & (z<100):\n",
    "        max_magnitude = 3\n",
    "        if z > 3:\n",
    "            z = 3\n",
    "        if z < -3:\n",
    "            z = -3\n",
    "        # scale measurements from 0 to 1\n",
    "        coordinate = 1 - (z-(-max_magnitude))/(3-(-max_magnitude))\n",
    "\n",
    "        color = pc.sample_colorscale('rdbu', samplepoints=float(coordinate), low=0.0, high=1.0, colortype='rgb')\n",
    "\n",
    "        Color = color[0]\n",
    "        Color = Color[Color.find('(')+1:Color.find(')')]\n",
    "        Color = Color.split(\",\")\n",
    "        Color = [Color[i].strip() for i in np.arange(0,len(Color))]\n",
    "        Color = np.array(Color)\n",
    "    \n",
    "    # If there is no data, make the color gray\n",
    "    if np.isnan(z):\n",
    "        Color = np.array([128,128,128])\n",
    "        \n",
    "    AA_Color_DF = pd.DataFrame(index=AA_DF.loc[:,'Number'],columns=[1,2,3])\n",
    "    #AA_Color_DF.index = AA_DF.loc[:,'Number']\n",
    "    \n",
    "    color_options = px.colors.qualitative.Alphabet\n",
    "    color_options = color_options[0:len(AA_Color_DF)]\n",
    "    \n",
    "    color_counter=0\n",
    "    for index in AA_Color_DF.index:\n",
    "        AA_Color_DF.loc[index,:] = hex_to_rgb(color_options[color_counter][1:None])\n",
    "        color_counter = color_counter+1\n",
    "    \n",
    "\n",
    "    if z>=100:\n",
    "        Color = np.array([AA_Color_DF.loc[z,1],AA_Color_DF.loc[z,2],AA_Color_DF.loc[z,3]])\n",
    "    \n",
    "    Color = Color.astype(int)\n",
    "    \n",
    "    return(Color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2732c073",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign a RGB color array to each entry in a 2D array.\n",
    "#     Store the RGB color arrays in a 3D array where the RGB values go down the 3rd axis\n",
    "def CreateImage(array,AA_DF):\n",
    "    dimension = len(array[0,:])\n",
    "    RGB_Array = np.empty(shape = [dimension,dimension,3])\n",
    "    for i in np.arange(0,dimension):\n",
    "        for j in np.arange(0,dimension):\n",
    "            value = array[i,j]\n",
    "            Array_To_Store = np.asarray(GetColor(value,AA_DF),dtype=int)\n",
    "            RGB_Array[i,j,0:3] = Array_To_Store\n",
    "    return(RGB_Array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "72c17b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a square, 2D array from a 1D array\n",
    "def SquareArray(array):\n",
    "    dimension = int(len(array)**0.5)\n",
    "    Squared_Array = np.empty(shape = [dimension,dimension])\n",
    "    for i in np.arange(0,dimension**2):\n",
    "        row = int(np.floor(i/dimension))\n",
    "        column = int(i - dimension*row)\n",
    "        Squared_Array[row,column] = array[i]\n",
    "    return(Squared_Array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "725a2bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "Test = np.array([9,4,7,3,1,4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba73bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "NewArray = Reorder(Test)\n",
    "NewArray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76edcc03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create function to reorder an array by sum of neighboring points\n",
    "#     neighboring pairs are placed in order from smallest to largest sums\n",
    "def Reorder(array):\n",
    "    SumArray = np.array([])\n",
    "    for i in np.arange(0,len(array)/2):\n",
    "        i = int(i)\n",
    "        if not np.isnan(array[2*i]):\n",
    "            Sum = array[2*i] + array[2*i+1]\n",
    "            SumArray = np.append(SumArray,Sum)\n",
    "    SortedIndices = np.argsort(SumArray)\n",
    "    \n",
    "    ExpandedSorted = np.array([])\n",
    "    for j in np.arange(0,len(SortedIndices)):\n",
    "        ExpandedSorted = np.append(ExpandedSorted,2*SortedIndices[j])\n",
    "        ExpandedSorted = np.append(ExpandedSorted,2*SortedIndices[j]+1)\n",
    "        \n",
    "    ExpandedSorted = ExpandedSorted.astype('int')\n",
    "    Replacements = array[ExpandedSorted]\n",
    "    array[0:len(Replacements)] = Replacements\n",
    "    \n",
    "    return(array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a75d5875",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hex_to_rgb(hex):\n",
    "  rgb = []\n",
    "  for i in (0, 2, 4):\n",
    "    decimal = int(hex[i:i+2], 16)\n",
    "    rgb.append(decimal)\n",
    "  \n",
    "  return(np.asarray(rgb))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
